lib/diff.ml

```diff
@@ -1,39 +1,271 @@
 open Types
 open Stddiff
+open Intermed
 
-type type_modification = {
-  type_kind : (type_decl_kind, type_kind) maybe_changed;
-  type_privacy : (Asttypes.private_flag, type_privacy) maybe_changed;
-  type_manifest : type_expr atomic_option;
-  type_params : (type_expr, type_param) list_;
-}
-
-and type_kind =
-  | Record_tk of (label_declaration, label) map
-  | Variant_tk of (Types.constructor_declaration, cstr_args) map
-  | Atomic_tk of type_decl_kind atomic_modification
+let type_expr ~typing_env ?(ref_params = []) ?(cur_params = []) ~reference
+    ~current () =
+  let normed_ref, normed_cur =
+    Normalize.params_arity ~reference:ref_params ~current:cur_params
+  in
+  if
+    Ctype.is_equal typing_env true
+      (List.map (fun param -> param.TypeDecl.type_expr) normed_ref
+      @ [ reference ])
+      (List.map (fun param -> param.TypeDecl.type_expr) normed_cur @ [ current ])
+  then None
+  else Some { reference; current }
 
-and label = {
-  label_type : type_expr maybe_changed_atomic;
-  label_mutable : (Asttypes.mutable_flag, field_mutability) maybe_changed;
-}
+module TypeDecl = struct
+  module TD = TypeDecl
+  module K = TD.Kind
+  module F = TD.Field
+  module C = TD.Constructor
 
-and field_mutability = Added_m | Removed_m
+  module Field = struct
+    type mutable_change = Added | Removed
 
-and cstr_args =
-  | Record_cstr of (label_declaration, label) map
-  | Tuple_cstr of type_expr maybe_changed_atomic_entry list
-  | Atomic_cstr of Types.constructor_arguments Stddiff.atomic_modification
+    type t = {
+      mutable_ : (bool, mutable_change) maybe_changed;
+      type_ : type_expr maybe_changed_atomic;
+    }
 
-and type_privacy = Added_p | Removed_p
-and type_param = (type_expr, type_param_diff) maybe_changed
-and type_param_diff = Added_tp of type_expr | Removed_tp of type_expr
+    let mutable_ ~reference ~current =
+      match (reference, current) with
+      | true, true | false, false -> Same reference
+      | true, false -> Changed Removed
+      | false, true -> Changed Added
 
-type type_ = {
-  tname : string;
-  tdiff : (type_declaration, type_modification) entry;
-}
+    let field ~typing_env ~ref_params ~cur_params reference current =
+      let mutable_ =
+        mutable_ ~reference:reference.F.mutable_ ~current:current.F.mutable_
+      in
+      let type_ =
+        type_expr ~typing_env ~ref_params ~cur_params
+          ~reference:reference.F.type_ ~current:current.F.type_ ()
+      in
+      match (mutable_, type_) with
+      | Same _, None -> None
+      | _, None -> Some { mutable_; type_ = Same reference.type_ }
+      | _, Some type_change -> Some { mutable_; type_ = Changed type_change }
+  end
+
+  module Constructor = struct
+    type args =
+      | Record of (F.t, Field.t) map
+      | Tuple of type_expr maybe_changed_atomic_entry list
+      | Unshared of C.args atomic_modification
```

Comment:
I think we've been going for `Atomic` for these, it would be best to use the same convention everywhere!